qiita_db/data.py

```diff
@@ -308,6 +329,65 @@ def create(cls, filetype, studies, filepaths=None):
 
         return rd
 
+    @classmethod
+    def delete(cls, raw_data_id, study_id):
+        """Removes the raw data with id raw_data_id
+
+        Parameters
+        ----------
+        raw_data_id : int
+            The raw data id
+        study_id : int
+            The study id
+
+        Raises
+        ------
+        ValueError
+            If the raw data id doesn't exist
+            If the raw data is not linked to that study_id
+            If the raw data has prep templates associated
+        """
+        conn_handler = SQLConnectionHandler()
+
+        # check if the raw data exist
+        if not cls.exists(raw_data_id):
+            raise ValueError("Raw data: %s doesn't exist." % str(raw_data_id))
+
+        study_raw_data_exists = conn_handler.execute_fetchone(
+            "SELECT EXISTS(SELECT * FROM qiita.study_raw_data WHERE "
+            "study_id = {0} AND raw_data_id = {1})".format(study_id,
+                                                           raw_data_id))[0]
+        if not study_raw_data_exists:
+            raise ValueError(
+                "Raw data %s is not linked to study %s" % (raw_data_id,
+                                                           study_id))
+
+        # check if there are any prep templates
+        prep_template_count = conn_handler.execute_fetchone(
+            "SELECT COUNT(*) FROM qiita.prep_template WHERE "
+            "raw_data_id = {0}".format(raw_data_id))[0]
+
+        if (prep_template_count > 0):
+            raise ValueError("Raw data: %s, has %d prep templates associated "
+                             "so it can't be erased." % (str(raw_data_id),
+                                                         prep_template_count))
+
+        # delete
+        conn_handler.execute("DELETE FROM qiita.study_raw_data WHERE "
+                             "raw_data_id = {0} AND "
+                             "study_id = {1}".format(raw_data_id, study_id))
+
+        # delete the connecting tables if there is no other study linked to
+        # the raw data
+        study_raw_data_count = conn_handler.execute_fetchone(
+            "SELECT COUNT(*) FROM qiita.study_raw_data WHERE "
+            "raw_data_id = {0}".format(raw_data_id))[0]
+        if study_raw_data_count == 0:
+            conn_handler.execute("DELETE FROM qiita.raw_filepath WHERE "
```

Comment:
That's correct. I didn't require the raw data to have no linked files because then you will not be able to erase a raw data from a second study without unlinking. One option, if you feel strongly about it, is to not allow erase it's the last raw data (remove all FK), if it has linked files.